File: scripts/compare_i18n.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Helper to read and eval file content to get the object
// We can't directly import because of potential environment issues with Vue/Webpack specific syntax if any
// But these files look like pure JS objects.
// However, they use 'export default', so we can strip that and eval.

function loadLangFile(filePath) {
    const content = fs.readFileSync(filePath, 'utf8');
    // Remove "export default" and any leading comments/whitespace
    const objectContent = content.replace(/export default\s*/, '').replace(/\/\/.*/g, '');
    // We need to make it a valid expression to eval. 
    // Note: eval is dangerous but we trust these files.
    // We might need to handle 'require' if used.
    const require = () => 'DUMMY_REQUIRE'; // Mock require
    try {
        return eval('(' + objectContent + ')');
    } catch (e) {
        console.error(`Error parsing ${filePath}:`, e);
        return {};
    }
}

const zhPath = path.resolve(__dirname, '../web/src/lang/zh.js');
const enPath = path.resolve(__dirname, '../web/src/lang/en.js');

const zh = loadLangFile(zhPath);
const en = loadLangFile(enPath);

function getKeys(obj, prefix = '') {
  let keys = [];
  for (const key in obj) {
    const newPrefix = prefix ? `${prefix}.${key}` : key;
    // If it's an object and not an array, recurse
    if (typeof obj[key] === 'object' && obj[key] !== null && !Array.isArray(obj[key])) {
        // Check if it's a leaf node (some keys might be objects but treated as values in some contexts? 
        // usually i18n files are nested objects until string)
        // But wait, if one is string and other is object, that's a mismatch too.
        keys.push(newPrefix); // Add the object key itself too
        keys = keys.concat(getKeys(obj[key], newPrefix));
    } else {
        keys.push(newPrefix);
    }
  }
  return keys;
}

const zhKeys = new Set(getKeys(zh));
const enKeys = new Set(getKeys(en));

const missingInEn = [...zhKeys].filter(k => !enKeys.has(k));
const missingInZh = [...enKeys].filter(k => !zhKeys.has(k));

console.log('Missing in EN:', missingInEn);
console.log('Missing in ZH:', missingInZh);

File: scripts/compare_i18n.py
import re
import os

def extract_keys(file_path):
    keys = set()
    stack = []
    
    with open(file_path, 'r', encoding='utf-8') as f:
        lines = f.readlines()
        
    for line in lines:
        line = line.strip()
        # Remove comments
        if '//' in line:
            line = line.split('//')[0].strip()
            
        if not line:
            continue
            
        # Match key: {
        match_obj_start = re.match(r'([a-zA-Z0-9_]+):\s*{', line)
        if match_obj_start:
            key = match_obj_start.group(1)
            stack.append(key)
            current_path = '.'.join(stack)
            keys.add(current_path)
            continue
            
        # Match key: 'value' or key: "value" or key: value
        match_val = re.match(r'([a-zA-Z0-9_]+):\s*', line)
        if match_val and '{' not in line:
            key = match_val.group(1)
            current_path = '.'.join(stack + [key])
            keys.add(current_path)
            
        # Match closing brace
        if line.startswith('}'):
            if stack:
                stack.pop()
                
    return keys

zh_path = 'web/src/lang/zh.js'
en_path = 'web/src/lang/en.js'

zh_keys = extract_keys(zh_path)
en_keys = extract_keys(en_path)

missing_in_en = zh_keys - en_keys
missing_in_zh = en_keys - zh_keys

print("Missing in EN:")
for k in sorted(missing_in_en):
    print(k)
    
print("\nMissing in ZH:")
for k in sorted(missing_in_zh):
    print(k)
